operator/duties/scheduler.go

```diff
@@ -0,0 +1,341 @@
+}
+
+type Scheduler struct {
+	beaconNode          BeaconNode
+	network             networkconfig.NetworkConfig
+	validatorController ValidatorController
+	slotTicker          SlotTicker
+	executeDuty         ExecuteDutyFunc
+
+	handlers            []dutyHandler
+	blockPropagateDelay time.Duration
+
+	reorg      chan ReorgEvent
+	indicesChg chan bool
+	ticker     chan phase0.Slot
+	waitCond   *sync.Cond
+
+	lastBlockEpoch            phase0.Epoch
+	currentDutyDependentRoot  phase0.Root
+	previousDutyDependentRoot phase0.Root
+	builderProposals          bool
+}
+
+func NewScheduler(opts *SchedulerOptions) *Scheduler {
+	s := &Scheduler{
+		beaconNode:          opts.BeaconNode,
+		network:             opts.Network,
+		slotTicker:          opts.Ticker,
+		executeDuty:         opts.ExecuteDuty,
+		validatorController: opts.ValidatorController,
+		builderProposals:    opts.BuilderProposals,
+		indicesChg:          opts.IndicesChg,
+		blockPropagateDelay: blockPropagateDelay,
+		ticker:              make(chan phase0.Slot),
+	}
+	s.handlers = []dutyHandler{
+		NewAttesterHandler(),
+		NewProposerHandler(),
+		NewSyncCommitteeHandler(),
+	}
+	if s.builderProposals {
+		s.handlers = append(s.handlers, NewValidatorRegistrationHandler())
+	}
+	return s
+}
+
+type ReorgEvent struct {
+	Slot     phase0.Slot
+	Previous bool
+	Current  bool
+}
+
+func (s *Scheduler) Run(ctx context.Context, logger *zap.Logger, ready chan<- struct{}) error {
+	logger = logger.Named(logging.NameDutyScheduler)
+	logger.Info("starting scheduler")
+
+	// Initialize the reorg channel
+	s.reorg = make(chan ReorgEvent)
+
+	s.waitCond = sync.NewCond(&sync.Mutex{})
+
+	// Subscribe to head events.  This allows us to go early for attestations if a block arrives, as well as
+	// re-request duties if there is a change in beacon block.
+	if err := s.beaconNode.Events(ctx, []string{"head"}, s.HandleHeadEvent(logger)); err != nil {
+		return fmt.Errorf("failed to subscribe to head events: %w", err)
+	}
+
+	handlerPool := pool.New().WithContext(ctx)
+	var reorgChs []chan<- ReorgEvent
+	var indicesChangeChs []chan<- bool
+	for _, handler := range s.handlers {
+		h := handler
+		slotTicker := make(chan phase0.Slot)
+		s.slotTicker.Subscribe(slotTicker)
+
+		indicesChangeCh := make(chan bool)
+		indicesChangeChs = append(indicesChangeChs, indicesChangeCh)
+		reorgCh := make(chan ReorgEvent)
+		reorgChs = append(reorgChs, reorgCh)
+		h.Setup(s.beaconNode, s.network, s.validatorController, s.ExecuteDuties, slotTicker, reorgCh, indicesChangeCh)
+
+		handlerPool.Go(func(ctx context.Context) error {
+			h.HandleDuties(ctx, logger)
+			return nil
+		})
+	}
+
+	s.slotTicker.Subscribe(s.ticker)
+	go s.SlotTicker(ctx, logger)
+
+	go fanOut(ctx, s.indicesChg, indicesChangeChs)
+	go fanOut(ctx, s.reorg, reorgChs)
+
+	if ready != nil {
+		ready <- struct{}{}
+	}
+	return handlerPool.Wait()
+}
+
+// TODO: use a fan out package or Prysm's event.Feed, because sends in this implementation are blocking
```

Comment:
two approaches can be taken, 
- in, and subscribers channels should be buffered
- add a default clause to the select.

I guess the first one is the best because the default will mean we will skip events and that will cause issues.

Reply:
we could also fall into default and retry only the subscribers that were blocking, still I think a small buffer will be ok since none of these event should happen a lot.

Reply:
the problem with buffering is that if you were stuck for say 12 seconds (the max timeout), then after 12 sec after the slot has already advanced, you resume reading from slotTicker, but then you get the previous slot from the ticker bc it was buffered

the default clause may work, we have this prototype:
```go
func fanOut[T any](ctx context.Context, in <-chan T, subscribers []chan<- T) {
    for {
        select {
        case <-ctx.Done():
            return
        case item := <-in:
            outbox := make([]chan<- T, len(subscribers))
            copy(outbox, subscribers)

            for len(outbox) > 0 {
                for i, subscriber := range outbox {
                    // Fan out the message to all subscribers.
                    select {
                    case <-ctx.Done():
                        return
                    case subscriber <- item:
                        // Sent -> Remove the subscriber from the outbox.
                        outbox = append(outbox[:i], outbox[i+1:]...)
                    default:
                    }
                }
            }
        }
    }
}
```

however bc it gets a bit complex, we're considering just using Prysm's event.Feed which does a similar thing already